# Проект по SQL - Анализ базы данных сервиса для чтения книг по подписке

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

___

**Цель исследования:** анализ базы данных приобретённого сервиса для чтения книг по подписке для формулировки ценностного предложение для нового продукта.

**Исходные данные:** база данных с информацией о книгах, издательствах, авторах, а также пользовательских обзоров книг.

<a id="fp_sql_documentation"></a>
**Описание данных:**

<u>**(1) Данные о книгах — books</u>** 📖

* **`book_id`** — идентификатор книги;
* **`author_id`** — идентификатор автора;
* **`title`** — название книги;
* **`num_pages`** — количество страниц;
* **`publication_date`** — дата публикации книги;
* **`publisher_id`** — идентификатор издателя.

<u>**(2) Данные об авторах — authors</u>** 👨

* **`author_id`** — идентификатор автора;
* **`author`** — имя автора.

<u>**(3) Данные об издательствах — publishers</u>** 🏢

* **`publisher_id`** — идентификатор издательства;
* **`publisher`** — название издательства.

<u>**(4) Данные о пользовательских оценках книг — ratings</u>** ⭐

* **`rating_id`** — идентификатор оценки;
* **`book_id`** — идентификатор книги;
* **`username`** — имя пользователя, оставившего оценку;
* **`rating`** — оценка книги.

<u>**(5) Данные о пользовательских обзорах — reviews</u>** 🔍

* **`review_id`** — идентификатор обзора;
* **`book_id`** — идентификатор книги;
* **`username`** — имя автора обзора;
* **`text`** — текст обзора.

___
**Схема данных:**

<img 
     src="other/images/scheme_sql.png" 
     height="800" 
     width="800" />

---
## ℹ️ Инструкция по выполнению проекта

* Опишите цели исследования;
* Исследуйте таблицы — выведите первые строки;
* Сделайте по одному SQL-запросу для решения каждого задания;
* Выведите результаты каждого запроса в тетрадке;
* Опишите выводы по каждой из решённых задач.

___
## Загрузка данных и ознакомление

**(1) Получим доступ к базе данных:**

In [1]:
# импортируем библиотеки

import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

**(2) Выведем первые строки таблиц:**

In [2]:
def data(sql_query):
    query = sql_query
    con=engine.connect()
    return pd.io.sql.read_sql(sql=text(query), con = con)

**`books`** - данные о книгах

In [3]:
data('''

SELECT * 
FROM books 
LIMIT 5

''')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
data('''

SELECT COUNT(*) 
FROM books

''')

count
0   1000

Всего в таблице **1 000 строк**.

---
**`authors`** - данные об авторах

In [5]:
data('''

SELECT * 
FROM authors 
LIMIT 5

''')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
data('''

SELECT COUNT(*) 
FROM authors

''')

count
0    636

Всего в таблице **636 строк**.

---
**`publishers`** - данные об издательствах

In [7]:
data('''

SELECT * 
FROM publishers 
LIMIT 5

''')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
data('''

SELECT COUNT(*) 
FROM publishers

''')

count
0    340

Всего в таблице **340 строк**.

---
**`ratings`** - данные о пользовательских оценках книг

In [9]:
data('''

SELECT * 
FROM ratings 
LIMIT 5

''')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
data('''

SELECT COUNT(*) 
FROM ratings

''')

count
0   6456

Всего в таблице **6 456 строк**.

---
**`reviews`** - данные о пользовательских обзорах

In [11]:
data('''

SELECT * 
FROM reviews 
LIMIT 5

''')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
data('''

SELECT COUNT(*) 
FROM reviews

''')

count
0   2793

Всего в таблице **2 793 строки**.

## Задание №1

---
**Посчитайте, сколько книг вышло после 1 января 2000 года.**
___

In [13]:
data('''

SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'

''')

count
0    819

---
**<u>Вывод:</u>** 

После 1 января 2000 года вышло **819 книг**.

## Задание №2

---
**Для каждой книги посчитайте количество обзоров и среднюю оценку.**
___

In [14]:
data('''

SELECT b.book_id, 
       b.title, 
       COUNT(DISTINCT rev.review_id) AS count_reviews, 
       ROUND(AVG(rat.rating), 2) AS avg_rating
FROM books AS b

LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
LEFT JOIN ratings AS rat ON b.book_id = rat.book_id

GROUP BY b.book_id, b.title
ORDER BY count_reviews DESC

''')

book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        963                                Water for Elephants   
2        734                                   The Glass Castle   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...   
4        695  The Curious Incident of the Dog in the Night-Time   
..       ...                                                ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996      808                            The Natural Way to Draw   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites   
998      221                          Essential Tales and Poems   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     count_reviews  avg_rating  
0                7        3.66  
1                6        3.98  
2                6        4.21  
3                6        4.41  
4                6        4.08  
..             ...         ...  
995              0        3.67  
996              0        3.00  
997              0        5.00  
998              0        4.00  
999              0        4.00  

[1000 rows x 4 columns]

---
**<u>Вывод:</u>** 

Для каждой книги посчитаны количество обзоров и средняя оценка. Больше всего обзоров было сделано по книге **Twilight (Twilight #1)** - **7 обзоров**. Максимальный балл средней оценки - **5.00** (данную оценку получило большое количество книг, определённого лидера нет).

## Задание №3

---
**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.**
___

In [15]:
data('''

WITH t_1 AS (SELECT publisher_id, 
                    book_id
             FROM books
             WHERE num_pages > 50
)

SELECT t_1.publisher_id,
       p.publisher
FROM t_1 

JOIN publishers AS p ON t_1.publisher_id = p.publisher_id

GROUP BY t_1.publisher_id, p.publisher
ORDER BY COUNT(t_1.book_id) DESC
LIMIT 1

''')

publisher_id      publisher
0           212  Penguin Books

---
**<u>Вывод:</u>**

Издательство, которое выпустило наибольшее число книг толще 50 страниц - **Penguin Books**.

## Задание №4

---
**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.**
___

In [16]:
data('''

WITH t_1 AS (SELECT book_id
             FROM ratings AS r
             GROUP BY book_id
             HAVING COUNT(rating_id) > 50
)

SELECT a.author_id, 
       a.author, 
       ROUND(AVG(r.rating), 2) AS avg_rating
FROM books AS b

JOIN t_1 ON b.book_id = t_1.book_id
JOIN authors AS a ON b.author_id = a.author_id
JOIN ratings AS r ON t_1.book_id = r.book_id

GROUP BY a.author_id, a.author
ORDER BY AVG(r.rating) DESC
LIMIT 1

''')

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré        4.29

---
**<u>Вывод:</u>**

Автор(ы) с самой высокой средней оценкой книг (при учёте книг с 50 и более оценками) - **J.K. Rowling/Mary GrandPré** (средняя оценка - **4.29**).

## Задание №5

---
**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.**
___

In [17]:
data('''

WITH t_1 AS (SELECT rev.username,
                    COUNT(review_id) AS count_review
             FROM reviews AS rev
             WHERE username IN (SELECT rat.username
                                FROM ratings AS rat
                                GROUP BY username
                                HAVING COUNT(rating_id) > 48)
                                GROUP BY username)
                                
SELECT SUM(count_review) / COUNT(count_review) AS avg_reviews
FROM t_1

''')

avg_reviews
0         24.0

---
**<u>Вывод:</u>**

Пользователи, которые поставили больше 48 оценок в среднем делают **24 обзора**.

## Общий вывод

**(1)** После 1 января 2000 года вышло **819 книг**.

---
**(2)** Для каждой книги посчитаны количество обзоров и средняя оценка. Больше всего обзоров было сделано по книге **Twilight (Twilight #1)** - **7 обзоров**. Максимальный балл средней оценки - **5.00** (данную оценку получило большое количество книг, определённого лидера нет).

---
**(3)** Издательство, которое выпустило наибольшее число книг толще 50 страниц - **Penguin Books**.

---
**(4)** Автор(ы) с самой высокой средней оценкой книг (при учёте книг с 50 и более оценками) - **J.K. Rowling/Mary GrandPré** (средняя оценка - **4.29**).

---
**(5)** Пользователи, которые поставили больше 48 оценок в среднем делают **24 обзора**.